odoo/addons/base/models/res_config.py

```diff
@@ -636,14 +636,13 @@ def execute(self):
         self.set_values()
 
         # module fields: install/uninstall the selected modules
-        to_install = []
+        to_install = self.env['ir.module.module']
         to_uninstall_modules = self.env['ir.module.module']
-        lm = len('module_')
-        for name, module in classified['module']:
-            if int(self[name]):
-                if module.state == "installed":
-                    continue
-                to_install.append((name[lm:], module))
+        for module in classified['module']:
+            fname = f'module_{module.name}'
+            if self[fname]:
+                if module.state != 'installed':
+                    to_install += module
             else:
                 if module and module.state in ('installed', 'to upgrade'):
                     to_uninstall_modules += module
```

Comment:
In two lines...
```python
to_install = classified['module'].filtered(lambda m: self[f'module_{m.name}'] and m.state != 'installed')
to_uninstall = classified['module'].filtered(lambda m: not self[f'module_{m.name}'] and m.state in ('installed', 'to upgrade'))
```
Remember that `+=` on recordsets in a loop has O(n²) in time complexity.